internal/plugin.go

```diff
@@ -1621,6 +1627,79 @@ func (p *Plugin) createPlacement(
 	return
 }
 
+// generateSelector determines the type of input and creates a map of selectors to be used in either the
+// clusterSelector or labelSelector field
+func (p *Plugin) generateSelector(
+	resolvedSelectors map[string]interface{},
+) (map[string]interface{}, error) {
+	if resolvedSelectors == nil {
+		return map[string]interface{}{"matchExpressions": []interface{}{}}, nil
+	}
+
+	// Ensure legacy selectors and formatted YAML selectors are not both in use
+	selectorsUsed := map[string]bool{
+		"legacy":       false,
+		"fullSelector": false,
+	}
+
+	for key := range resolvedSelectors {
+		if key == "matchExpressions" || key == "matchLabels" {
+			selectorsUsed["fullSelector"] = true
+		} else {
+			selectorsUsed["legacy"] = true
+		}
+	}
+
+	if selectorsUsed["legacy"] && selectorsUsed["fullSelector"] {
+		return nil, fmt.Errorf("cannot use both key/value pairs and formatted YAML for placement selectors")
+	}
+
+	resolvedSelectorsJSON, err := json.Marshal(resolvedSelectors)
+	if err != nil {
+		return nil, err
+	}
+
+	resolvedSelectorsLS := metav1.LabelSelector{}
+	decoder := json.NewDecoder(bytes.NewReader(resolvedSelectorsJSON))
+	decoder.DisallowUnknownFields()
+
+	err = decoder.Decode(&resolvedSelectorsLS)
+	if err != nil {
+		resolvedSelectorsLS = metav1.LabelSelector{}
+
+		// Check if it's a legacy selector
+		for label, value := range resolvedSelectors {
+			valueStr, ok := value.(string)
+			if !ok {
+				return nil, fmt.Errorf(
+					"matchExpressions has non-string type as value: %s",
```

Comment:
I think if you use this error then the validation code from above that I commented on can be deleted.
```suggestion
					"the input is not a valid label selector or simple key-value label matching map",
```

Reply:
@mprahl this is true but isn't it better to have the error message be a bit more specific?

Reply:
If I just put:
```yaml
clusterSelector:
    cloud: some-cloud
    somethingelse: 123
```

I'd be confused if the error said `matchExpressions`.

Reply:
ok I'll change it